Various component tests and some conditional statements

tmashuang opened the pull request:
Conditional in account-menu in removeAccount when keyring sometimes is not initially provideed
Conditional on unlock-page when there is no target.getBoundingClientRect on the element.

tmashuang committed:
Various component tests and some conditional statements

Conditional in account-menu in removeAccount when keyring sometimes is not initially provideed
Conditional on unlock-page when there is no target.getBoundingClientRect on the element.

tmashuang committed:
Lint

tmashuang committed:
Update helpers

tmashuang committed:
Remove component debugging

tmashuang committed:
Lint

tmashuang commented on ui/app/pages/unlock-page/tests/unlock-page.test.js:
```diff
@@ -0,0 +1,70 @@
+
+  const store = configureMockStore()(mockStore)
+
+  stubComponent(Mascot)
```
The Mascot is initialized with the `unlock.component`, https://github.com/MetaMask/metamask-extension/blob/7b730ebc472a86c8e9b6f12f8136b20131fed544/ui/app/pages/unlock-page/unlock-page.component.js#L144-L148

The Mascot Component requires a DOM, one that append to and animates with the mouse and key presses. It seems that the enzyme doesn't know how to handle these. Which lead me to stub the react component and return `null` for it to prevent those animations while testing the `unlock.component` in isolation.

Gudahtt replied:
Ahhh I see, that makes sense. So this is just a means of stubbing it out.

I think this would be better accomplished using something like `proxyquire` to return a stub in place of that component - as this approach of stubbing the Mascot class itself will mutate that class globally for the duration of this Node.js session. This would break any tests on Mascot itself, for example, though it seems in this case it isn't used anywhere else, which is why it's working.

tmashuang replied:
I played around with `proxyquire` but could get it to play nicely with the router stuff. Will look into it more, and get more familiar with it since some of the component tests are stubbed with it.

Gudahtt replied:
After looking into this a bit more, I think you're right that `proxyquire` doesn't handle this case well. `proxyquire` is only meant to stub out dependencies of the module you're importing directly, whereas this is one level deeper than that (as `Mascot` is being imported by the component, which itself is being imported by the container which is what's being imported here).

Though... looking at these tests, they are only really testing the component. One solution would be to import it directly, and test that instead. I managed to get that to work here: https://github.com/Gudahtt/metamask-extension/commit/27711c7347f5cee9f30785850a4b8f5278f1f570

Though another solution might be to just fix Mascot, and stop mutating the DOM. React makes it quite easy to achieve the same ends without doing that. I'll give that a try.

Gudahtt replied:
The Mascot has been updated in #7893. I've confirmed that once that is merged, the component stubbing done here isn't required anymore.

tmashuang commented on ui/app/components/app/account-menu/account-menu.component.js:
```diff
@@ -193,6 +193,12 @@ export default class AccountMenu extends Component {
   renderRemoveAccount (keyring, identity) {
     const { t } = this.context
+
+    // Sometimes keyrings aren't loaded yet
```
Without it, in the `account-menu.test.js`, the component will render with 
>TypeError: Cannot destructure property `type` of 'undefined' or 'null'.

which refers to `state.metamask.keyrings`

After this change the component will render correctly with the props. This was the only way to get it to render and do basic tests on it. I just copied that lines from `removeAccount()`, https://github.com/MetaMask/metamask-extension/blob/7b730ebc472a86c8e9b6f12f8136b20131fed544/ui/app/components/app/account-menu/account-menu.component.js#L229-L232 since it seems that it doesn't initially load.

tmashuang commented on ui/app/components/app/account-menu/tests/account-menu.test.js:
```diff
@@ -0,0 +1,208 @@
+  })
+
+  afterEach(() => {
+    props.toggleAccountMenu.resetHistory()
```
For my own reference: 
> Since sinon@5.0.0, the sinon object is a default sandbox. Unless you have a very advanced setup or need a special configuration, you probably want to only use that one.

Will change the behaviors of the `resetHistory/restore/etc` to get a more streamlied approach.

tmashuang committed:
Rebase with develop and adjust tests
Add default params for render helpers

tmashuang committed:
Remove stubComponent for old Mascot
Changes in https://github.com/MetaMask/metamask-extension/pull/7893 has prevented the need to stub it out.

Change logout to lock in account-menu test

Gudahtt reviewed (approved):
Thanks, LGTM!
